### ISSUE TITLE
SAML response is valid using www.samltool.com but fails on php-saml

### ISSUE THREAD

**Issue by marsoltys** [onelogin/php-saml#167]
Hi,

I'm using Laravel with aacotroneo/laravel-saml2 package and php-saml toolkit (2.10)

When I take base64 encoded saml response and use SAML Tool website to 
1. Base64 DECODE
2. Decrypt SAML Response
3. Validate Saml Response

Resposne seems to be valid (and even without timing issues).
I've taken the same decoded and decrypted response inside of  OneLogin_Saml2_Auth using Xdebug and it still passes through www.samltool.com

But when I'm trying to sign in using saml-php it fails giving "Reference validation failed".

In Xdebug it looks like xml is changed.
Script adds "default" to XML so it looks like `< default:Assertion>...</default:Assertion>`
Obviously as this XML is passed as a string to hash function it will give different hash result.

Any idea if this something on my end or problem with php-saml toolkit ?
Am I missing something ?

Let me know if you need more details.

**Comment by pitbulk** [comment]
www.samltool.com uses internally php-saml 2.9.1 (will be soon updated to 2.10.0).

Any modification on the SAMLResponse, even a simple extra space, will produce a "Reference validation failed". Try to use directly the demo of this php-saml toolkit and see if you experience the same issue, if not, maybe the problem is at lavarel integration or in your environment.

**Comment by marsoltys** [comment]
I'll test that but I believe it will give me the same results. I have copied and tested Response in saml tool during my app debug, after Laravel handled it and it has been validated successfully but not in the app itself.

I'll get you my debug data so you can see what I mean. 

Another weird  behavior was that when on saml tool I entered encrypted, base64 encoded response and filled in private key I had "Reference validation failed" as well. Only when I've followed steps as in my first comment, response was validated.

**Comment by pitbulk** [comment]
Try with php-saml 2.10.0 and 2.9.1 version.

**Comment by marsoltys** [comment]
I've tried both and I'm getting the same error.
I've found that when SAML response is not encrypted it is validated successfully.

So currently validation fails only when response is encrypted.
In that case response is decrypted correctly in php-saml but fails on signature/digest validation.

When I'll add code below in xmlseclibs.php on line 861:
`$data = str_replace(":default", "", str_replace("default:", "", $data));`
Response is validated properly (!?)

As in my first message xmlseclibs adds "default" namespace to Assertion node and its child nodes (I'm not sure why)

Any ideas ?
Thanks M

**Comment by pitbulk** [comment]
This seems a namespace issue. I think xmlseclibs adds the "default" namespace when the SAMLResponse does not contain it on the Assertion element.

If I'm not wrong, the signature is inside the Assertion, that is encrypted at the SAMLResponse.

In this case, the toolkit follow those steps:
- Base64 decode the SAMLResponse parameter to get the XML.
- Check if there is an EncryptedAssertion.
- If there is a Message signed, it validated the Signature with the original SAMLResponse.
- If there is a signature on the decrypted Assertion, it validates it.

I think that something is wrong on the original Assertion that was encrypted, something related with the namespaces that makes that the signature validation fails.

But if you think that the problem is at xmlseclibs, I recommend you to open an issue at  [xmlseclibs lib repository](https://github.com/robrichards/xmlseclibs/issues) and there in order to let robrichards to debug I recommend.
- Generate a new pair of private key/certificate at https://www.samltool.com/self_signed_certs.php and use it on php-saml
- Upload the new public cert of php-saml at ADFS
- Generate a SAMLResponse encrypted that return the Reference validation failed and attach it to the issue you opened.

**Comment by pitbulk** [comment]
This user mentioned that experience a similar [issue](https://github.com/onelogin/php-saml/issues/168) than yours and says that the signature validation pass on v2.8.0. Can you try that version?

**Comment by marsoltys** [comment]
True. It is passing validation on v2.8.0.

**Comment by pitbulk** [comment]
Ok, I think that the issue is related to the namespace...on 2.9.0 [I removed](https://github.com/onelogin/php-saml/commit/39878dd041bc7fa91f19e9c04ef4e4008e3079cc) the EncryptedAssertion node that encapsulates the Assertion element, but it seems that if the Assertion has no namespace.. the signature process fail.

Can you provide an example of the SAMLResponse (base64encoded)  and the public and private key of the SP in order to debug myself and try to fix the issue?

**Comment by pitbulk** [comment]
With that patch https://github.com/onelogin/php-saml/commit/5434c9d8484af5233a58a93ee758bf4539c6134e @sitecode solved the issue...

but we must use the same decrypted document for extract data than for validate signature (in order to avoid signature wrapping attacks).

**Comment by marsoltys** [comment]
Emailed you @gmail

Thanks

**Comment by pitbulk** [comment]
Got it, thanks

**Comment by sblommaert** [comment]
I can confirm this, just started having the same issue. I can also confirm downgrading to 2.8.0 works. Don't think I know enough to help out, but I'll be happy to test a patch if you have one.

**Comment by marsoltys** [comment]
Just to let you know that patch 5434c9d8484af5233a58a93ee758bf4539c6134e doesn't work for me

**Comment by pitbulk** [comment]
@marsoltys , @sblommaert , @nagargaurav

91c2510 should fix the issue, please test the php-saml master branch.

**Comment by sblommaert** [comment]
Patch seems to work for me! Nice! Are you creating a new release for this fix?

**Comment by pitbulk** [comment]
I will create a new release.

**Comment by marsoltys** [comment]
Confirmed. It is working.

**Comment by pitbulk** [comment]
Released https://github.com/onelogin/php-saml/releases/tag/v2.10.1

**Comment by scriptted** [comment]
Hi,

I still have this issue using the latest release. I had to downgrade to 2.8, maybe some specific cases are not covered by the patch, feel free to ask for any detail that might be the cause of the regression.

**Comment by pitbulk** [comment]
Can you provide the SAMLResponse that is failing to you + private key to decrypt it?
(If is a production server, maybe you can temporally replace the private key/certificate used in order to be able to provide it)

**Comment by scriptted** [comment]
Here you go !
[response.txt](https://github.com/onelogin/php-saml/files/580973/response.txt)
[key.txt](https://github.com/onelogin/php-saml/files/580974/key.txt)

**Comment by pitbulk** [comment]
Is that response.txt correct? I'm not able to base64 decode it.

**Comment by scriptted** [comment]
I just decoded it on samltool with success, what soft did you open it with ?

**Comment by pitbulk** [comment]
When I
1. [Base64 DECODE](https://www.samltool.com/base64.php)
2. [Decrypt SAML Response](https://www.samltool.com/decrypt.php)
3. [Validate Saml Response](https://www.samltool.com/validate_response.php)

I get "Reference validation failed" so not the same issue described on that ticket.
Let me review what's going on with that response.

**Comment by scriptted** [comment]
It's validated when I test it on samltool

**Comment by pitbulk** [comment]
![invalid_response](https://cloud.githubusercontent.com/assets/600897/20143414/38b30424-a699-11e6-9874-ae7db59db56a.png)

**Comment by scriptted** [comment]
You entered the wrong X.509 cert, you must enter the one for the IdP, you will find it attached 
[testshib_cert.txt](https://github.com/onelogin/php-saml/files/581141/testshib_cert.txt)

**Comment by pitbulk** [comment]
I saved your response at tests/data/responses/scriptted.xml.base64
and this test Scriptted167Test.php at tests/src/Onelogin/Saml2/
```
<?php


class OneLogin_Saml2_Scriptted167Test extends PHPUnit_Framework_TestCase
{

    public function testScriptted167()
    {
        $settingsDir = TEST_ROOT .'/settings/';
        include $settingsDir.'settings1.php';
        $settingsInfo['strict'] = false;
        $settingsInfo['idp']['x509cert'] = 'MIIDAzCCAeugAwIBAgIVAPX0G6LuoXnKS0Muei006mVSBXbvMA0GCSqGSIb3DQEBCwUAMBsxGTAXBgNVBAMMEGlkcC50ZXN0c2hpYi5vcmcwHhcNMTYwODIzMjEyMDU0WhcNMzYwODIzMjEyMDU0WjAbMRkwFwYDVQQDDBBpZHAudGVzdHNoaWIub3JnMIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAg9C4J2DiRTEhJAWzPt1S3ryhm3M2P3hPpwJwvt2q948vdTUxhhvNMuc3M3S4WNh6JYBs53R+YmjqJAII4ShMGNEmlGnSVfHorex7IxikpuDPKV3SNf28mCAZbQrX+hWA+ann/uifVzqXktOjs6DdzdBnxoVhniXgC8WCJwKcx6JO/hHsH1rG/0DSDeZFpTTcZHj4S9MlLNUtt5JxRzV/MmmB3ObaX0CMqsSWUOQeE4nylSlp5RWHCnx70cs9kwz5WrflnbnzCeHU2sdbNotBEeTHot6a2cj/pXlRJIgPsrL/4VSicPZcGYMJMPoLTJ8mdy6mpR6nbCmP7dVbCIm/DQIDAQABoz4wPDAdBgNVHQ4EFgQUUfaDa2mPi24x09yWp1OFXmZ2GPswGwYDVR0RBBQwEoIQaWRwLnRlc3RzaGliLm9yZzANBgkqhkiG9w0BAQsFAAOCAQEASKKgqTxhqBzROZ1eVy++si+eTTUQZU4+8UywSKLia2RattaAPMAcXUjO+3cYOQXLVASdlJtt+8QPdRkfp8SiJemHPXC8BES83pogJPYEGJsKo19l4XFJHPnPy+Dsn3mlJyOfAa8RyWBS80u5lrvAcr2TJXt9fXgkYs7BOCigxtZoR8flceGRlAZ4p5FPPxQR6NDYb645jtOTMVr3zgfjP6Wh2dt+2p04LG7ENJn8/gEwtXVuXCsPoSCDx9Y0QmyXTJNdV1aB0AhORkWPlFYwp+zOyOIR+3m1+pqWFpn0eT/HrxpdKa74FA3R2kq4R7dXe4G0kUgXTdqXMLRKhDgdmA==';
        $settingsInfo['sp']['privateKey'] = 'MIIEvwIBADANBgkqhkiG9w0BAQEFAASCBKkwggSlAgEAAoIBAQDOsjHtMtkU+Ponz8GyCeNdCuu+ccTZXyziHHV8Ecm/bLOP6pqIUm5RWXh6AEp5ANWEZLwU5oS4wHAsLPZKOXZLPMTPvPZi07X8MQ4PSPmnm/EItYl9QPbKoe5VrxTCGfMwE9ktwOe4kab4uk5cjWxuQja8Rug7Fxz6BYXasYz3moIhIBWEmOdnDSvEpU/QikaVnr05nvahQhz4H0zL/xrhTvAgKVYjQ+kd1im52grJ9yvIt/GE44A2NI8Pc+FHMIFdlJQ9KqFxFRMM325SZJj3fidQKt/UdnQ4Vx9OYZqdo/n2Z6icR4SRb157Jokr85gLPiXi9TGnh18B15xj7pIpAgMBAAECggEBAJlavJAwWb+lunQvQJbHn1k48Q+reVGNu89HLiOPTxAnplHwzha2kgLpTum1rYt5gvFni+Q14+LvZEpiUZlgaNQJkjxPBr7sNOmXu8a7bf9dQ8giaI/10yUkX6BS9iPM/RF+qHO53DKkXi3MjLc+dmSX2+R093hf/a/CyT0o3l4VfyWc9HQ0GZLWf1pg2uN1ICzv+VHCP8ZWVPzjNLf5gCKYksIWWSCRR2R8mX9/CS/t10AWeBZNC8LvNhpLxo7BwaOwizrLfr4b+zCec1nt4BeIj8NMnEkYmwpzIVAZOxo2dicoc0GAlj5YomtIAFXPkquKKa46HQ+Ot+zGAWsEC8ECgYEA8Luu2c1ueXBRNzqKjh4nMLb2W64xL6kyltrvJf+xLcMN4hWBWr8Sgh1fQ1lNX3IkZJB5LHcNnT4P0+WARF6r/6HV+Gb+SYaaDGGOekyuuqHu+wv+ytG4ARsNDld5JCu90U0efHt/zH/spxVMByBqRxM2MmGr7IeJhgpwffHIBuMCgYEA283rEq90fJWw/iNVMLM6t1f1w1YLjCAgOkhb8m1glZ1+A+B9NtPA/JQc8a+FIFr8xnjzwTIzCxEavlCbaTb73sm3dETpPVSQdTUp2IQ60/HGkDHTzXZgliQ03ExwR6ywabSqVBa8TlrjVWh+ivxfvf+2okFUVH1Ziydt86XChIMCgYAMIQiwRJ96UuRkqfhtLKvA0pcbC23C/v1jiGhgO7/LlHZQm9DGp7XhBGXWWNuDIZcf0f+o896lu4jEnl2WFEcKPDcxqs7JGZUNZqTtGIv9L0eBMpuF0Ic61wgTXujln6aSsxYEI5zOyus7h+EqwsFYD2m9pBorcrDFJiWA0vvHDQKBgQCqftCPnt85YBd5GckTzIKIAseUbCLq32SAMO55XlQj2ZUMxs35gEla4jyNT/twfINEJD/XJOwo8myPIRFbjJXUWx1dJ7LPqs+eys0ZsH+9NPYzlSLYZbFH5O/HzdbuEcqJQ+U28qfAE4nxDuvRPS+6viZZN1mFa/HQGZaU/oeipwKBgQDgn2Nixnks4syN6tQ/vvM/Bg3dfp/NaKGXttydtGCbVYkitulOGaexh3fCZUkA63js6/Hgch9XGVyvJYLN1wMLDGpuEMN7rDKjBL1f34XI9boSAQzh79hafZ5Mcnomd6tatXE8O3m8UIJiH7iedbnw225URjXSiQ9tPW9N066g3A==';
        $settingsInfo['sp']['x509cert'] = null;

        $settings = new OneLogin_Saml2_Settings($settingsInfo);

        $xml = file_get_contents(TEST_ROOT . '/data/responses/scriptted.xml.base64');

        $response = new OneLogin_Saml2_Response($settings, $xml);
        $this->assertTrue($response->isValid());
    }
}
```

and it seems that the response validation pass on 2.10.0 but no in 2.10.1.

**Comment by pitbulk** [comment]
@scriptted, https://github.com/onelogin/php-saml/commit/994cacb85e419a289b6fff727eb068109e7f4591 should fix the issue you experimented

**Comment by scriptted** [comment]
Nice, thanks for the fix !

**Comment by BrandonNoad** [comment]
Any idea when the next release is going to be?

**Comment by pitbulk** [comment]
I plan to release this with PR #175 this week.

**Comment by pitbulk** [comment]
Released [php-saml 2.10-2](https://github.com/onelogin/php-saml/releases/tag/v2.10.2)
